File: src/game/gameplay/RaceTimer.js
export class RaceTimer {
  constructor(checkpointManager) {
    this.elapsedTime = 0;
    this.hasStarted = false;
    this.isRunning = false;

    checkpointManager.onCheckpointPassed((nextCheckpointIndex) => {
      if (!this.hasStarted && nextCheckpointIndex > 0) {
        this.start();
      }
    });

    checkpointManager.onCourseFinished(() => {
      this.stop();
    });
  }

  update(dt) {
    if (!this.isRunning) {
      return;
    }

    this.elapsedTime += dt;
  }

  start() {
    this.hasStarted = true;
    this.isRunning = true;
  }

  stop() {
    this.isRunning = false;
  }

  reset() {
    this.elapsedTime = 0;
    this.hasStarted = false;
    this.isRunning = false;
  }

  getFormattedTime() {
    const minutes = Math.floor(this.elapsedTime / 60);
    const seconds = this.elapsedTime % 60;
    const secondsText = seconds.toFixed(2).padStart(5, "0");
    return `${String(minutes).padStart(2, "0")}:${secondsText}`;
  }
}

File: Assets/_Project/Scripts/Gameplay/RaceTimer.cs
using UnityEngine;

/// <summary>
/// Tracks the elapsed time of the race.
/// Starts on first checkpoint pass and ends on course finish.
/// </summary>
public class RaceTimer : MonoBehaviour
{
    [SerializeField] private CheckpointManager checkpointManager;
    [SerializeField] private bool autoStartOnFirstCheckpoint = true;

    private bool isRunning;
    private bool hasStarted;
    private float elapsedTime;

    public bool IsRunning => isRunning;
    public bool HasStarted => hasStarted;
    public float ElapsedTime => elapsedTime;

    public void SetCheckpointManager(CheckpointManager manager)
    {
        if (checkpointManager == manager)
        {
            return;
        }

        if (isActiveAndEnabled && checkpointManager != null)
        {
            checkpointManager.OnCheckpointPassed -= HandleCheckpointPassed;
            checkpointManager.OnCourseFinished -= HandleCourseFinished;
        }

        checkpointManager = manager;

        if (isActiveAndEnabled && checkpointManager != null)
        {
            checkpointManager.OnCheckpointPassed += HandleCheckpointPassed;
            checkpointManager.OnCourseFinished += HandleCourseFinished;
        }
    }

    private void OnEnable()
    {
        if (checkpointManager == null)
        {
            return;
        }

        checkpointManager.OnCheckpointPassed += HandleCheckpointPassed;
        checkpointManager.OnCourseFinished += HandleCourseFinished;
    }

    private void OnDisable()
    {
        if (checkpointManager == null)
        {
            return;
        }

        checkpointManager.OnCheckpointPassed -= HandleCheckpointPassed;
        checkpointManager.OnCourseFinished -= HandleCourseFinished;
    }

    private void Update()
    {
        if (!isRunning)
        {
            return;
        }

        elapsedTime += Time.deltaTime;
    }

    private void HandleCheckpointPassed(int nextCheckpointIndex, int totalCheckpoints)
    {
        if (!autoStartOnFirstCheckpoint)
        {
            return;
        }

        if (!hasStarted && nextCheckpointIndex > 0)
        {
            StartTimer();
        }
    }

    private void HandleCourseFinished()
    {
        StopTimer();
    }

    public void StartTimer()
    {
        hasStarted = true;
        isRunning = true;
    }

    public void StopTimer()
    {
        isRunning = false;
    }

    public void ResetTimer()
    {
        isRunning = false;
        hasStarted = false;
        elapsedTime = 0f;
    }

    public string GetFormattedTime()
    {
        int minutes = Mathf.FloorToInt(elapsedTime / 60f);
        float seconds = elapsedTime % 60f;
        return string.Format("{0:00}:{1:00.00}", minutes, seconds);
    }
}
